cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -0,0 +1,632 @@
+    auto in_data = in.GetValues<uint64_t>(1);
+    auto out_data = out->GetMutableValues<out_type>(1);
+    auto ymd_caster = make_ymd_caster(in.type);
+    for (int64_t i = 0; i < in.length; i++) {
+      out_data[i] = static_cast<unsigned>(ymd_caster(in_data[i]).month());
+    }
+    return Status::OK();
+  }
+};
+
+// ----------------------------------------------------------------------
+// Extract day from timestamp
+
+template <typename out_type>
+struct Day {
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    const int64_t& in_data = internal::UnboxScalar<const TimestampType>::Unbox(in);
+    auto ymd_caster = make_ymd_caster(in.type);
+    checked_cast<Int64Scalar*>(out)->value =
+        static_cast<unsigned>(ymd_caster(in_data).day());
+    return Status::OK();
+  }
+
+  static Status Call(KernelContext* ctx, const ArrayData& in, ArrayData* out) {
+    auto in_data = in.GetValues<uint64_t>(1);
+    auto out_data = out->GetMutableValues<out_type>(1);
+    auto ymd_caster = make_ymd_caster(in.type);
+    for (int64_t i = 0; i < in.length; i++) {
+      out_data[i] = static_cast<unsigned>(ymd_caster(in_data[i]).day());
+    }
+    return Status::OK();
+  }
+};
+
+// ----------------------------------------------------------------------
+// Extract week from timestamp
+
+template <typename out_type>
+struct Week {
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    const auto& in_data = internal::UnboxScalar<const TimestampType>::Unbox(in);
+    auto week_caster = make_week_caster(in.type);
+    checked_cast<Int64Scalar*>(out)->value = week_caster(in_data);
+    return Status::OK();
+  }
+
+  static Status Call(KernelContext* ctx, const ArrayData& in, ArrayData* out) {
+    auto in_data = in.GetValues<uint64_t>(1);
+    auto out_data = out->GetMutableValues<out_type>(1);
+    auto week_caster = make_week_caster(in.type);
+    for (int64_t i = 0; i < in.length; i++) {
+      out_data[i] = week_caster(in_data[i]);
+    }
+    return Status::OK();
+  }
+};
+
+// ----------------------------------------------------------------------
+// Extract quarter from timestamp
+
+template <typename out_type>
+struct Quarter {
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    const auto& in_data = internal::UnboxScalar<const TimestampType>::Unbox(in);
+    auto ymd_caster = make_ymd_caster(in.type);
+    checked_cast<Int64Scalar*>(out)->value =
+        (static_cast<unsigned>(ymd_caster(in_data).month()) - 1) / 3 + 1;
+    return Status::OK();
+  }
+
+  static Status Call(KernelContext* ctx, const ArrayData& in, ArrayData* out) {
+    auto in_data = in.GetValues<uint64_t>(1);
+    auto out_data = out->GetMutableValues<out_type>(1);
+    auto ymd_caster = make_ymd_caster(in.type);
+    for (int64_t i = 0; i < in.length; i++) {
+      out_data[i] = (static_cast<unsigned>(ymd_caster(in_data[i]).month()) - 1) / 3 + 1;
+    }
+    return Status::OK();
+  }
+};
```

Comment:
You're repeating yourself a lot. You basically have specific extraction function (the "caster" functions) and a generic kernel structure that just applies the extraction function. It should be possible to minimize the amount of code copied and pasted in this file.

Reply:
Indeed. If I "dry" this up do you feel the approach is ok?

Reply:
I've abstracted the "caster maker" function. I'll try to do the same with the generic kernel.